api/helpers/add-game-to-match.js

```diff
@@ -0,0 +1,73 @@
+const dayjs = require('dayjs');
+
+module.exports = {
+  friendlyName: 'Add Game to Match',
+
+  description:
+    'Adds the provided game to the appropriate match. Will create a match if not found and will end the match if this game finishes it.',
+
+  inputs: {
+    game: {
+      type: 'ref',
+      description:
+        'Game record to be added to the appropriate match. Should include an array of players',
+      required: true,
+    },
+  },
+  fn: async ({ game }, exits) => {
+    try {
+      if (!game.ranked) {
+        return exits.success();
+      }
+      const [player1, player2] = game.players;
+      let relevantMatch = await sails.helpers.findOrCreateCurrentMatch(player1.id, player2.id);
+      if (!relevantMatch) {
+        return exits.error(new Error('Could not add game to match'));
+      }
+      if (relevantMatch.endTime || relevantMatch.winner) {
+        // Set game back to unranked if the player match already finished
+        await Game.updateOne(game.id).set({ ranked: false });
+        return exits.success();
+      }
+      if (!relevantMatch.games) {
+        relevantMatch.games = [];
+      }
+      let numPlayer1Wins = 0;
+      let numPlayer2Wins = 0;
+
+      for (const priorGame of [game, ...relevantMatch.games]) {
+        if (priorGame.result === 0) {
+          if (priorGame.p0 === relevantMatch.player1) {
+            numPlayer1Wins++;
+          } else {
+            numPlayer2Wins++;
+          }
+        } else if (priorGame.result === 1) {
+          if (priorGame.p1 === relevantMatch.player1) {
+            numPlayer1Wins++;
+          } else {
+            numPlayer2Wins++;
+          }
+        }
+      }
+      // Add game to match
+      await Match.addToCollection(relevantMatch.id, 'games').members([game.id]);
+
+      // End match if this game clinches it
+      if (numPlayer1Wins >= 2) {
+        relevantMatch = await Match.updateOne(relevantMatch.id).set({
+          endTime: dayjs().valueOf(),
+          winner: game.players[0].id,
+        });
+      } else if (numPlayer2Wins >= 2) {
+        relevantMatch = await Match.updateOne(relevantMatch).set({
+          endTime: dayjs().valueOf(),
+          winner: game.players[1].id,
+        });
+      }
```

Comment:
Since we always do one or the other, does it make sense to limit how much we're repeating?
```suggestion
    let winner;
    if (numPlayer1Wins >= 2) {
      winner = game.players[0].id;
    } else if (numPlayer2Wins >= 2) {
      winner = game.players[1].id;
    }

    relevantMatch = await Match.updateOne(relevantMatch).set({
      endTime: dayjs().valueOf(),
      winner,
    });
```

Reply:
The funky bit is that we only need to update the `relevantMatch` if either player wins (in that case we set the `winner` and `endTime` fields), but if this is say the 2nd game and the players are tied 1-1, then we don't actually want to set an `endTime` or a winner. We could set it up so that we set variables (defaulting to `null`) for winner and endTime and then always update the match at the end (with possibly null values), but this does result in unnecessary queries for games where the match doesn't actually need an update b/c it's not over yet.

Thoughts?